### PR TITLE
Dangling bond detection for ReactionPattern

### PR DESCRIPTION
**Pull request by alubbock**
A dangling bond is when a specified bond has no binding partner, e.g.

```python
A(a=None) % A(a=1)
```

Patterns with dangling bonds are disallowed in Kappa and pending deprecation in BioNetGen. This PR adds dangling bond detection when `ReactionPattern` patterns, allowing users to detect these mistakes early (before network generation).

### CODE REVIEW

File: pysb/pattern.py [hunk 1]
```diff
@@ -1,25 +1,27 @@
...
-    ReactionPattern, ANY, as_complex_pattern
+    ReactionPattern, ANY, as_complex_pattern, DanglingBondError
 import networkx as nx
 from networkx.algorithms.isomorphism.vf2userfunc import GraphMatcher
 from networkx.algorithms.isomorphism import categorical_node_match
 import numpy as np
 
 
-def get_bonds_in_pattern(pat):
+def get_bonds_in_pattern(pat, as_set=True):
```
Comment (jmuhlich) [inline pysb/pattern.py]: This new interface seems unwieldy. A suggestion: Rename this function to `get_half_bonds_in_pattern(pat)` and have it always return the `list`. Then implement `get_bonds_in_pattern` as:
```python
def get_bonds_in_pattern(pat):
    return set(get_half_bonds_in_pattern(pat))
```
Then `check_dangling_bonds` would call `get_half_bonds_in_pattern`.

File: pysb/pattern.py [hunk 2]
```diff
@@ -63,7 +57,42 @@ def _get_bonds_in_monomer_pattern(mp):
...
+
+    Returns
+    -------
+    set
+        Bond numbers used in the supplied pattern
+
+    Examples
+    --------
+
+    These examples convert the set to a list for testing purposes only (sets
```
Comment (jmuhlich) [inline pysb/pattern.py]: Oof, this is unfortunate. What about writing the examples slightly differently to sidestep the issue? Maybe with equality tests:
```python
>>> get_bonds_in_pattern(A(b1=None, b2=None)) == set()
True
>>> get_bonds_in_pattern(A(b1=1) % A(b2=1)) == {1}
True
...
```

**Comment by alubbock** [comment]
Updated doctests as requested.

**Comment by bgyori** [comment]
Clarification question here: is this also meant to raise an error for a ReactionPattern like
```
A(b=1) + B(a=1)
```
which is technically not "dangling"?

**Comment by lh64** [comment]
Those would be considered dangling bonds. Both reactants in that case have a bond ‘1’ that doesn’t have a binding partner. So that should result in an error.

**Comment by alubbock** [comment]
The pattern `A(b=1) + B(a=1)` does indeed raise an error now in PySB master, as it should -- I agree with Leonard's logic, because bonds cannot span across separate `ComplexPattern`s.

**Pull request merged**